### PR TITLE
Fix #698 by adjusting the regex to not have \.  Still not 100% sure w…

### LINKED ISSUE #698: Invalid block names confuse the library
On Blockly Dev Tools, create a block in the library named '123'. It will be saved as '_\123' which generates JSON with '_\123'. If you then export the library and reimport it you sometimes get an error, other times you just get a confused library that has the blocks listed under the unadjusted name (123) and offers to save your new block (_\123).

The two issues here:
- Invalid names should just get an underscore prefix. The backslash is not needed.
- When importing, invalid block names should get adjusted as well.

### PR DESCRIPTION
**Pull request by picklesrus**
…hy that was there.

Also replaces bad names on input.  There are probably more invalid names but this is
a start.

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/700)

### CODE REVIEW

**Comment by RoboErikG** [comment]
:lgtm: 

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[demos/blockfactory/app_controller.js, line 204 at r1](https://reviewable.io:443/reviews/google/blockly/700#-KUXzrrhktoMHVKlElzZ:-KUXzrrhktoMHVKlElz_:b-355hu2) ([raw file](https://github.com/google/blockly/blob/482aba088842fc2e66248b2f7b8f466b1022fd70/demos/blockfactory/app_controller.js#L204)):*
> ```JavaScript
>     // Some names are invalid so fix them up.
>     blockType = FactoryUtils.cleanBlockType(blockType);
>     
> ```

nit: ws

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/700#-:-KUXzuOBvlM_2HVWLsQ9:bg7igkv)*

**Comment by RoboErikG** [comment]
Was there anything else that we're waiting on for this?

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/700#-:-KVX0XcfyH3-6KfGE3aE:b-90seag)*

**Comment by picklesrus** [comment]
Just for me to context-switch... There was some discussion that we didn't need to disallow most of what this (and the existing regex does.  I just haven't had a chance to think it through, undo it,  and then test. 

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/700#-:-KVeZW_uh7bmX8YKK4uK:bu26334)*

File: demos/blockfactory/factory_utils.js [hunk 1]
```diff
@@ -57,6 +57,19 @@ FactoryUtils.getBlockDefinition = function(blockType, rootBlock, format, workspa
 };
 
 /**
+ * Convert invalid block name to a valid one. Replaces whitespace
+ * and prepend names that start with a digit with an '_'.
+ * @param {string} blockType Type of block
+ * @return {string} cleaned up block type.
```
Comment (NeilFraser) [inline demos/blockfactory/factory_utils.js]: End comment with period (param line), start comment with uppercase (return line).

File: demos/blockfactory/block_library_controller.js [hunk 2]
```diff
@@ -64,8 +64,8 @@ BlockLibraryController = function(blockLibraryName, opt_blockLibraryStorage) {
 BlockLibraryController.prototype.getCurrentBlockType = function() {
   var rootBlock = FactoryUtils.getRootBlock(BlockFactory.mainWorkspace);
   var blockType = rootBlock.getFieldValue('NAME').trim().toLowerCase();
-  // Replace white space with underscores
-  return blockType.replace(/\W/g, '_').replace(/^(\d)/, '_\\1');
+  // Replace invalid characters
```
Comment (rachel-fenichel) [inline demos/blockfactory/block_library_controller.js]: Period at end of comment then lgtm.

**Comment by picklesrus** [comment]
Fixed up comments and I'm going to merge this since it does solve a real problem, even if it doesn't fix the fact that the regex is more restrictive than it needs to be. I'll add a comment in https://github.com/google/blockly/issues/656 which is also about the regex being to restrictive.

**Pull request merged**